File: lc-archived/java/109.convert-sorted-list-to-binary-search-tree.java
/*
 * @lc app=leetcode id=109 lang=java
 *
 * [109] Convert Sorted List to Binary Search Tree
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if(head == null)return null;
        if(head.next == null)return new TreeNode(head.val);
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode slow = dummy, fast = head;
        while(fast != null&&fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode root = slow.next;
        slow.next = null;
        TreeNode left = sortedListToBST(head);
        TreeNode right = sortedListToBST(root.next);
        return new TreeNode(root.val, left, right);
    }
}
// @lc code=end

File: lc-archived/cpp/109.convert-sorted-list-to-binary-search-tree.cpp
/*
 * @lc app=leetcode id=109 lang=cpp
 *
 * [109] Convert Sorted List to Binary Search Tree
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// When it comes to tree or linked-list,
// just think about recursion and solve it recursively.

class SolutionA
{
private:
    TreeNode *helper(ListNode *beg, ListNode *end)
    {
        if (beg == nullptr || beg == end)
            return nullptr;
        int len = 0;
        ListNode *cur = beg;
        for (len = 0, cur = beg; cur != end; ++len)
            cur = cur->next;
        int mid = len / 2;
        for (cur = beg; mid != 0; --mid)
            cur = cur->next;
        TreeNode *res = new TreeNode(cur->val);
        res->left = helper(beg, cur);
        res->right = helper(cur->next, end);
        return res;
    }

public:
    TreeNode *sortedListToBST(ListNode *head)
    {
        return helper(head, nullptr);
    }
};

class Solution
{
    // Using fast and slow pointers to find middle pointer.
private:
    TreeNode *helper(ListNode *beg, ListNode *end)
    {
        if (beg == nullptr || beg == end)
            return nullptr;
        int len = 0;
        ListNode *slow = beg, *fast = beg;
        while (fast != end && fast->next != end)
        {
            fast = fast->next->next;
            slow = slow->next;
        }
        TreeNode *res = new TreeNode(slow->val);
        res->left = helper(beg, slow);
        res->right = helper(slow->next, end);
        return res;
    }

public:
    TreeNode *sortedListToBST(ListNode *head)
    {
        return helper(head, nullptr);
    }
};
